### PR TITLE
Strict function types 180206625

### PR DESCRIPTION
**Pull request by Blodir**
Enables the `strictFunctionTypes` tsc rule.

There's a couple of `any`'s related to third party libraries that I didn't know how to type...

### CODE REVIEW

**Review by aorin — CHANGES_REQUESTED** [review]
Looks good! There is one place that I think could be improved. (Also `projects/laji/src/app/shared-modules/select/select-subcategories/select-subcategories.component.ts` is hard to understand but it's not really related to this pull request)

File: projects/laji/src/app/+taxonomy/taxon/taxon.component.ts [hunk 1]
```diff
@@ -110,6 +112,7 @@ export class TaxonComponent implements OnInit, OnDestroy {
 
   private initTaxon(taxonId: string): Observable<any> {
```
[the remark this answers is not in the document]
Comment (Blodir) [reply projects/laji/src/app/+taxonomy/taxon/taxon.component.ts]: True, fixed :)

**Review by aorin — APPROVED** [review]
:+1:

File: projects/laji/src/app/shared-modules/select/select-subcategories/select-subcategories.component.ts [hunk 2]
```diff
@@ -158,7 +158,9 @@ export class SelectSubcategoriesComponent implements OnChanges {
       }
       let tmpSelectedOptions = [];
       let tmpUnselectedOptions = [];
-      const countObj = this.selectedOptions[i].filter((item: SelectOptions) => item.id !== undefined).length;
+      const countObj = this.selectedOptions[i].filter(
+        (item: SelectOptions | string) => typeof item === 'string' || item.id !== undefined
+      ).length;
```
Comment (olzraiti) [inline projects/laji/src/app/shared-modules/select/select-subcategories/select-subcategories.component.ts:163]: BTW this broke the component, see task https://www.pivotaltracker.com/story/show/180599328

```typeof item === 'string' || item.id !== undefined```

is not the same as:
```item.id !== undefined```

I fixed this with:
```typeof item !== 'string' && item.id !== undefined```

Comment (olzraiti) [reply projects/laji/src/app/shared-modules/select/select-subcategories/select-subcategories.component.ts:163]: Hmm, it should be the same though. Maybe some other changes here affecting this. I pushed the above as a hotfix to prod anyways - let's have a better look next week.

**Pull request merged**